### PR TITLE
feat: support generating cpu partitioning file from infra flag

### PR DESCRIPTION
**Pull request by eggfoobar**
/hold
Needs PR: openshift/api#1284
EP: [openshift/enhancement#1213](https://github.com/openshift/enhancements/pull/1213)

Signed-off-by: ehila <ehila@redhat.com>

**- What I did**

Add support for generating a machine config file that will configure kubelet for workload partitioning based off of the infrastructure flag.

**- How to verify it**
When the Infrastructure status resource is set to `cpuPartitioning: AllNodes`, we will generate two new MCs for workload partitioning named `01-master-cpu-partitioning` and `01-worker-cpu-partitioning` that will create a configuration file for kubelet to use on start up.
```yaml
apiVersion: config.openshift.io/v1
kind: Infrastructure
metadata:
  name: cluster
status:
  cpuPartitioning: AllNodes
```

**- Description for the changelog**

Add new generated MCs for workload partitioning based off of infrastructure status.

### CODE REVIEW

**Comment by cgwalters** [comment]
There's prior art for the installer generating MachineConfigs based on the install config, see e.g. https://github.com/openshift/installer/blob/master/pkg/asset/machines/machineconfig/hyperthreading.go

I personally don't have a strong opinion on this, but I would say that it makes sense to have the "install time configuration to MachineConfiguration" logic live mainly in one place, whether that's github.com/openshift/installer or here.

File: pkg/controller/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -183,6 +188,14 @@ func (b *Bootstrap) Run(destDir string) error {
 		configs = append(configs, kconfigs...)
 	}
 
+	if infraConfig.Status.CPUPartitioning != apicfgv1.CPUPartitioningNone {
```
Comment (yuqi-zhang) [inline pkg/controller/bootstrap/bootstrap.go:191]: Discussed on slack, but to clarify, if this only applies to bootstrap, if the MC does get generated, the install will fail due to a mismatch of configs generated by bootstrap and regular, causing a "can't find rendered-master-xxx" when the install happens. Basically https://github.com/openshift/machine-config-operator/issues/2114

**Comment by eggfoobar** [comment]
Thanks folks, appreciate the feedback, I'll close this PR for now.

File: pkg/controller/bootstrap/bootstrap.go (hunk 1 above)
Comment (eggfoobar) [reply pkg/controller/bootstrap/bootstrap.go:191]: Hey @yuqi-zhang , was looking through the code a bit, trying to understand this a little better. Does this mean that it's not possible to have a MC generated and maintain by MCO during bootstrap?

Comment (yuqi-zhang) [reply pkg/controller/bootstrap/bootstrap.go:191]: No, it is possible. You must have both the bootstrap and the "in-cluster" flow generate the same configs.

As an example, see this: https://github.com/openshift/machine-config-operator/pull/3015

which adds both for nodes.config.openshift.io objects

**Comment by eggfoobar** [comment]
/reopen 

Re-opening this PR, after some discussion with the installer team, MCO seems to be the best place to enforce the CPU Partitioning configuration cluster wide.

**Comment by eggfoobar** [comment]
New PR openshift/machine-config-operator#3355

**Pull request closed without merging**